Move selector naming to the descriptor

emilsjolander opened the pull request:
Currently two selector names are hard-coded in DOM.java.
"<this_element>" and "Accessibility Properties". With this commit
descriptors are able to themselves choose what the selector name should
be and how their various properties should be distributed among the
selectors.

This enables an element to have styles such as:

```
layout {
  width: 100px;
  height: 100px;
}

props {
  text: "Hello World";
}
```

Due to this requiring an added property in StyleAccumulator I chose to
also take this commit to split up styles from computed styles which are
two separate concepts but up until now have been handled the same. A
style may be `width: 100%` while a computed style would be `width:
320px`.

rickbrew commented on stetho/src/main/java/com/facebook/stetho/inspector/elements/android/DialogFragmentDescriptor.java:
```diff
@@ -24,6 +24,7 @@
 import com.facebook.stetho.inspector.elements.DescriptorMap;
 import com.facebook.stetho.inspector.elements.NodeType;
 import com.facebook.stetho.inspector.elements.StyleAccumulator;
+import com.facebook.stetho.inspector.elements.ComputedStyleAccumulator;
```
nit: to keep things nice and alphabetical, this should move up a few lines

(and there's a few more of these in the diff, I'll just leave this 1 comment for them)

rickbrew commented on stetho/src/main/java/com/facebook/stetho/inspector/protocol/module/CSS.java:
```diff
@@ -28,6 +29,8 @@
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
+import java.util.HashMap;
```
nit: alphabetical ordering of this too :)

rickbrew commented on stetho/src/main/java/com/facebook/stetho/inspector/protocol/module/CSS.java:
```diff
@@ -96,30 +96,7 @@ public JsonRpcResult getMatchedStylesForNode(JsonRpcPeer peer, JSONObject params
-    match.rule = rule;
-
-    rule.style.shorthandEntries = Collections.emptyList();
+    final Map<String, List<CSSProperty>> selectors = new HashMap<>();
```
using a `HashMap` for temporary storage like this is actually pretty heavy weight ... is there an alternative? how many keys to we expect to have here? can we reuse/recycle, especially since the strings we put in here will probably be repeated over and over?

for instance: maybe we create and cache 1 `HashMap` in an `mSelectorsCache` field. during this method, we still use it as a way to keep track of the `ArrayList` for each selector. then, we `put(selector, null)` at the end instead of either throwing away the `HashMap` or `clear()`ing it. That way we don't constantly churn on the `HashMap.Entry` allocations

I think it's reasonable to do this in a follow-up PR as well, it doesn't have to block us right now

rickbrew replied:
and, we may as well type the `Map`'s `V`alue as `ArrayList` instead of `List` -- doing this can sometimes enable certain static optimizations

rickbrew commented on stetho/src/main/java/com/facebook/stetho/inspector/protocol/module/CSS.java:
```diff
@@ -136,57 +113,46 @@ public void run() {
-                }
-              }
-            });
+                  List<CSSProperty> properties = selectors.get(selector);
```
`ArrayList` instead of `List` plz (see comment above!)

rickbrew commented on stetho/src/main/java/com/facebook/stetho/inspector/protocol/module/CSS.java:
```diff
@@ -136,57 +113,46 @@ public void run() {
 
-    return result;
-  }
+    for (String selectorName : selectors.keySet()) {
```
Here's another reason i'd prefer to avoid `HashMap` ... iterating over this requires a temporary `Iterator` allocation. Preferable to avoid, although not dogmatically so.

emilsjolander commented:
Addressed most comments (still iterating keySet()) and fixed some bugs in my implementation which I somehow did not catch last time I was testing this.

rickbrew commented on stetho/src/main/java/com/facebook/stetho/inspector/protocol/module/CSS.java:
```diff
@@ -27,12 +28,15 @@
   private final ChromePeerManager mPeerManager;
   private final Document mDocument;
   private final ObjectMapper mObjectMapper;
+  private final Map<String, ArrayList<CSSProperty>> mSelectorCollector = new HashMap<>();
```
please put a javadoc on this explaining how this is used and the restrictions on it -- certainly don't want a future maintainer using it on another thread, or not realizing that there is a "protocol" attached with correctly using it!

rickbrew commented on stetho/src/main/java/com/facebook/stetho/inspector/elements/android/ViewDescriptor.java:
```diff
@@ -42,6 +43,9 @@
   private static final String NONE_MAPPING = "<no mapping>";
+  private static final String VIEW_SELECTOR_NAME = "<this_view>";
+  private static final String ACCESSIBILITY_SELECTOR_NAME = "Accessibility Properties";
+
```
nit: an extra line of whitespace somehow appeared :)

rickbrew commented on stetho/src/main/java/com/facebook/stetho/inspector/protocol/module/CSS.java:
```diff
@@ -136,57 +112,56 @@ public void run() {
-    return result;
-  }
+    for (String selectorName : mSelectorCollector.keySet()) {
+      final ArrayList<CSSProperty> properties = mSelectorCollector.get(selectorName);
```
is there any reason we're iterating over the `keySet()` and then retrieving the value via `get()`, versus just iterating over the `entrySet()` and using `Map.Entry#getKey()` and `#getValue()` ? they're functionally equivalent, although the latter (iterating over entrySet) avoids an extra hashtable lookup ... which isn't very expensive ... 

iterating the `keySet()` would make perfect sense if we're using an implementation of `Map` that creates the `Map.Entry`s on-demand, such as `ArrayMap` or (iirc) `IdentityHashMap`, but `HashMap` creates them up-front

rickbrew commented on stetho/src/main/java/com/facebook/stetho/inspector/protocol/module/CSS.java:
```diff
@@ -136,57 +119,56 @@ public void run() {
 
-    return result;
-  }
+    for (Map.Entry<String, ArrayList<CSSProperty>> enrty : mSelectorCollector.entrySet()) {
```
typo: `enrty` should be `entry`

rickbrew commented:
ok looks good. builds locally, and don't really want to wait for Travis-CI
